File: FacadePattern/cpp/src/office_device.cpp
#include <iostream>
#include "header/office_device.h"

bool Light::getIsOn(){
    return isOn;
}

int Light::getBrightness(){
    return brightness;
}

void Light::on(){
    if(!isOn){
        isOn = true;
        cout << "Light is on." << endl;
    }
}

void Light::off(){
    if(isOn){
        isOn = false;
        cout << "Light is off." << endl;
    }
}

void Light::brightnessUp(){
    if(brightness >= Light_MAX_LEVEL || ++brightness == Light_MAX_LEVEL){
        cout << "The current brightness is maximum (level " << Light_MAX_LEVEL << ")." << endl;
    }else{
        cout << "The brightness is up (level " << brightness << ")." << endl;;
    }
}

void Light::brightnessDown(){
    if(brightness <= 1 || --brightness == 1){
        cout << "The current brightness is minimum (level 1)." << endl;
    }else{
        cout << "The brightness is down (level " << brightness << ")." << endl;
    }
}


bool AirConditioner::getIsOn(){
    return isOn;
}

int AirConditioner::getTemperature(){
    return temperature;
}

void AirConditioner::on(){
    if(!isOn){
        isOn = true;
        cout << "Air Conditioner is on." << endl;
    }
}

void AirConditioner::off(){
    if(isOn){
        isOn = false;
        cout << "Air Conditioner is off." << endl;
    }
}

void AirConditioner::temperatureUp(){
    if(temperature >= AIR_CONDITIONER_MAX_TEMPERATURE || ++temperature == AIR_CONDITIONER_MAX_TEMPERATURE){
        cout <<"The temperature is maximum (" << AIR_CONDITIONER_MAX_TEMPERATURE << " Celsius)." << endl;
    }else{
        cout << "The temperature is up (" << temperature << " Celsius)." << endl;
    }
}

void AirConditioner::temperatureDown(){
    if(temperature <= AIR_CONDITIONER_MIN_TEMPERATURE || --temperature == AIR_CONDITIONER_MIN_TEMPERATURE){
        cout <<"The temperature is minimum (" << AIR_CONDITIONER_MIN_TEMPERATURE << " Celsius)." << endl;
    }else{
        cout << "The temperature is down (" << temperature << " Celsius)." << endl;
    }
}


bool WaterDispenser::getIsOn(){
    return isOn;
}

void WaterDispenser::on(){
    if(!isOn){
        isOn = true;
        cout << "Water Dispenser is on." << endl;
    }
}

void WaterDispenser::off(){
    if(isOn){
        isOn = false;
        cout << "Water Dispenser is off." << endl;
    }
}


OfficeFacade::OfficeFacade(shared_ptr<Light> &light, shared_ptr<AirConditioner> &airConditioner, shared_ptr<WaterDispenser> &waterDispenser){
    this->light = light;
    this->airConditioner = airConditioner;
    this->waterDispenser = waterDispenser;
}

OfficeFacade::OfficeFacade(shared_ptr<Light> &&light, shared_ptr<AirConditioner> &&airConditioner, shared_ptr<WaterDispenser> &&waterDispenser){
    this->light = light;
    this->airConditioner = airConditioner;
    this->waterDispenser = waterDispenser;
}

bool OfficeFacade::isAllDeviceOn(){
    return light->getIsOn() && airConditioner->getIsOn() && waterDispenser->getIsOn();
}

void OfficeFacade::on(){
    light->on();
    airConditioner->on();
    waterDispenser->on();
}

void OfficeFacade::off(){
    light->off();
    airConditioner->off();
    waterDispenser->off();
}

int OfficeFacade::getBrightness(){
    return light->getBrightness();
}

int OfficeFacade::getTemperature(){
    return airConditioner->getTemperature();
}

void OfficeFacade::brightnessUp(){
    light->brightnessUp();
}

void OfficeFacade::brightnessDown(){
    light->brightnessDown();
}

void OfficeFacade::temperatureUp(){
    airConditioner->temperatureUp();
}

void OfficeFacade::temperatureDown(){
    airConditioner->temperatureDown();
}

File: FacadePattern/python/OfficeDevice.py
class Light:
    _MAX_LEVEL:int = 5
    def __init__(self) -> None:
        self._isOn:bool = False
        self._brightness:int = 1

    @property
    def isOn(self) -> bool:
        return self._isOn()
    
    @property
    def brightness(self) -> int:
        return self._brightness
    
    def on(self):
        if not self._isOn:
            self._isOn = True
            print("Light is on.")

    def off(self):
        if self._isOn:
            self._isOn = False
            print("Light is off.")

    def brightnessUp(self):
        if self._brightness < Light._MAX_LEVEL:
            self._brightness += 1

        if self._brightness == Light._MAX_LEVEL:
            print(f"The brightness is maximum (level {Light._MAX_LEVEL}).")
        else:
            print(f"The brightness is up (level {self._brightness}).")

    def brightnessDown(self):
        if self._brightness > 1:
            self._brightness -= 1

        if self._brightness <= 1:
            print("The brightness is minimum (level 1).")
        else:
            print(f"The brightness is down (level {self._brightness}).")


class AirConditioner:
    _MAX_TEMPERATURE:int = 30
    _MIN_TEMPERATURE:int = 16

    def __init__(self) -> None:
        self._isOn:bool = False
        self._temperature:int = AirConditioner._MAX_TEMPERATURE

    @property
    def isOn(self) -> bool:
        return self._isOn()
    
    @property
    def temperature(self) -> int:
        return self._temperature
    
    def on(self):
        if not self._isOn:
            self._isOn = True
            print("Air Conditioner is on.")

    def off(self):
        if self._isOn:
            self._isOn = False
            print("Air Conditioner is off.")

    def temperatureUp(self):
        if self._temperature < AirConditioner._MAX_TEMPERATURE:
            self._temperature += 1

        if self._temperature == AirConditioner._MAX_TEMPERATURE:
            print(f"The temperature is maximum ({AirConditioner._MAX_TEMPERATURE} Celsius).")
        else:
            print(f"The temperature is up ({self.temperature} Celsius).")

    def temperatureDown(self):
        if self._temperature > AirConditioner._MIN_TEMPERATURE:
            self._temperature -= 1

        if self._temperature == AirConditioner._MIN_TEMPERATURE:
            print(f"The temperature is minimum ({AirConditioner._MIN_TEMPERATURE} Celsius).")
        else:
            print(f"The temperature is down ({self.temperature} Celsius).")


class WaterDispenser:
    def __init__(self) -> None:
        self._isOn:bool = False

    @property
    def isOn(self) -> bool:
        return self._isOn()
    
    def on(self):
        if not self._isOn:
            self._isOn = True
            print("Water Dispenser is on.")

    def off(self):
        if self._isOn:
            self._isOn = False
            print("Water Dispenser is off.")


class OfficeFacade:
    def __init__(self, light:Light, airConditioner:AirConditioner, waterDispenser:WaterDispenser) -> None:
        self._light = light
        self._airConditioner = airConditioner
        self._waterDispenser = waterDispenser

    def isAllDeviceOn(self):
        return self._light.isOn and self._airConditioner.isOn and self._waterDispenser.isOn
    
    def on(self):
        self._light.on()
        self._airConditioner.on()
        self._waterDispenser.on()
        
    def off(self):
        self._light.off()
        self._airConditioner.off()
        self._waterDispenser.off()

    @property
    def brightness(self) -> int:
        return self._light.brightness
    
    @property
    def temperature(self) -> int:
        return self._airConditioner.temperature
    
    def brightnessUp(self):
        self._light.brightnessUp()

    def brightnessDown(self):
        self._light.brightnessDown()

    def temperatureUp(self):
        self._airConditioner.temperatureUp()

    def temperatureDown(self):
        self._airConditioner.temperatureDown()